File: socai/browser/cdp/endpoint.py
"""CDP endpoint discovery for a user-owned Chrome profile."""

from __future__ import annotations

import json
import os
import platform
import subprocess
import time
import urllib.error
import urllib.request
from dataclasses import asdict, dataclass
from pathlib import Path
from typing import Any


INSPECT_URL = "chrome://inspect/#remote-debugging"
DEFAULT_DEVTOOLS_PORTS = (9222, 9223)


@dataclass(frozen=True)
class Endpoint:
    source: str
    browser_ws_url: str
    http_version_url: str | None = None
    version: dict[str, Any] | None = None

    def to_dict(self) -> dict[str, Any]:
        return asdict(self)


def fetch_json(url: str, *, timeout: float = 5.0) -> dict[str, Any]:
    with urllib.request.urlopen(url, timeout=timeout) as response:
        return json.loads(response.read().decode("utf-8"))


def _try_fetch_json(url: str, *, timeout: float = 1.0) -> dict[str, Any] | None:
    try:
        return fetch_json(url, timeout=timeout)
    except (OSError, urllib.error.URLError, json.JSONDecodeError, TimeoutError):
        return None


def _endpoint_from_version_payload(version: dict[str, Any], *, source: str, version_url: str) -> Endpoint | None:
    browser_ws_url = str(version.get("webSocketDebuggerUrl") or "")
    if not browser_ws_url:
        return None
    return Endpoint(
        source=source,
        browser_ws_url=browser_ws_url,
        http_version_url=version_url,
        version={
            key: version.get(key)
            for key in ("Browser", "Protocol-Version", "User-Agent", "V8-Version")
            if key in version
        },
    )


def endpoint_from_http_url(url: str, *, source: str = "http_url", timeout: float = 5.0) -> Endpoint:
    version_url = f"{url.rstrip('/')}/json/version"
    try:
        version = fetch_json(version_url, timeout=timeout)
    except (OSError, urllib.error.URLError, json.JSONDecodeError, TimeoutError) as exc:
        raise RuntimeError(f"CDP HTTP endpoint is not reachable or did not expose /json/version: {url}") from exc

    endpoint = _endpoint_from_version_payload(version, source=source, version_url=version_url)
    if endpoint is None:
        raise RuntimeError(f"CDP HTTP endpoint did not return webSocketDebuggerUrl: {version_url}")
    return endpoint


def resolve_explicit_endpoint(
    *,
    browser_ws_url: str | None = None,
    http_url: str | None = None,
) -> Endpoint | None:
    """Resolve only explicitly supplied CDP endpoints."""

    if browser_ws_url:
        return Endpoint(source="argument", browser_ws_url=browser_ws_url)
    if http_url:
        return endpoint_from_http_url(http_url, source="argument")

    if env_ws := os.environ.get("SOCAI_CDP_WS"):
        return Endpoint(source="SOCAI_CDP_WS", browser_ws_url=env_ws)
    if env_http := os.environ.get("SOCAI_CDP_URL"):
        return endpoint_from_http_url(env_http, source="SOCAI_CDP_URL")

    return None


def resolve_cdp_endpoint(
    *,
    browser_ws_url: str | None = None,
    http_url: str | None = None,
) -> Endpoint:
    endpoint = resolve_explicit_endpoint(browser_ws_url=browser_ws_url, http_url=http_url)
    if endpoint is None:
        raise RuntimeError(
            "No CDP endpoint was provided. Pass browser_ws_url/Endpoint from the Socai app backend, "
            "or set SOCAI_CDP_WS / SOCAI_CDP_URL for local development."
        )
    return endpoint


def chrome_profile_roots() -> list[Path]:
    """Return profile roots where an already-open Chrome may expose CDP state."""

    if override := os.environ.get("SOCAI_CHROME_USER_DATA_DIR"):
        return [Path(override).expanduser()]

    home = Path.home()
    system = platform.system()
    if system == "Darwin":
        paths = [
            home / "Library/Application Support/Google/Chrome",
            home / "Library/Application Support/Comet",
            home / "Library/Application Support/Arc/User Data",
            home / "Library/Application Support/Microsoft Edge",
            home / "Library/Application Support/BraveSoftware/Brave-Browser",
        ]
    elif system == "Windows":
        paths = [
            home / "AppData/Local/Google/Chrome/User Data",
            home / "AppData/Local/Microsoft/Edge/User Data",
            home / "AppData/Local/BraveSoftware/Brave-Browser/User Data",
        ]
    else:
        paths = [
            home / ".config/google-chrome",
            home / ".config/chromium",
            home / ".config/microsoft-edge",
            home / ".config/BraveSoftware/Brave-Browser",
        ]

    seen: set[str] = set()
    result: list[Path] = []
    for path in paths:
        key = str(path)
        if key not in seen:
            seen.add(key)
            result.append(path)
    return result


def _endpoint_from_version_url(url: str, *, source: str) -> Endpoint | None:
    version = _try_fetch_json(url)
    if not version:
        return None
    return _endpoint_from_version_payload(version, source=source, version_url=url)


def _endpoint_from_active_port(profile_root: Path) -> Endpoint | None:
    marker = profile_root / "DevToolsActivePort"
    try:
        lines = marker.read_text(encoding="utf-8").splitlines()
    except (FileNotFoundError, NotADirectoryError, OSError):
        return None
    if not lines:
        return None

    try:
        port = int(lines[0].strip())
    except ValueError:
        return None
    ws_path = lines[1].strip() if len(lines) > 1 else ""

    try:
        return endpoint_from_http_url(f"http://127.0.0.1:{port}", source=f"active_port:{profile_root}")
    except RuntimeError as exc:
        cause = exc.__cause__
        if isinstance(cause, urllib.error.HTTPError) and cause.code == 404 and ws_path:
            return Endpoint(source=f"active_port:{profile_root}", browser_ws_url=f"ws://127.0.0.1:{port}{ws_path}")
        return None


def discover_existing_chrome_endpoint() -> Endpoint | None:
    """Find a CDP endpoint for an already-open browser profile."""

    explicit = resolve_explicit_endpoint()
    if explicit is not None:
        return explicit

    for profile_root in chrome_profile_roots():
        endpoint = _endpoint_from_active_port(profile_root)
        if endpoint is not None:
            return endpoint

    for port in DEFAULT_DEVTOOLS_PORTS:
        endpoint = _endpoint_from_version_url(f"http://127.0.0.1:{port}/json/version", source=f"port:{port}")
        if endpoint is not None:
            return endpoint
    return None


def wait_for_existing_chrome_endpoint(*, timeout: float = 45.0, poll_seconds: float = 0.5) -> Endpoint | None:
    deadline = time.time() + max(0.1, timeout)
    while time.time() < deadline:
        endpoint = discover_existing_chrome_endpoint()
        if endpoint is not None:
            return endpoint
        time.sleep(max(0.1, poll_seconds))
    return None


def open_remote_debugging_page() -> None:
    if platform.system() == "Darwin":
        subprocess.run(["open", "-a", "Google Chrome", INSPECT_URL], check=False)
        return

    import webbrowser

    webbrowser.open(INSPECT_URL, new=2)


discover_chrome_cdp = resolve_cdp_endpoint

File: core/src/cdp/endpoint.rs
use std::env;
use std::fs;
use std::path::{Path, PathBuf};
use std::time::{Duration, Instant};

use serde::{Deserialize, Serialize};

const INSPECT_URL: &str = "chrome://inspect/#remote-debugging";
const DEFAULT_DEVTOOLS_PORTS: &[u16] = &[9222, 9223];
const HTTP_TIMEOUT: Duration = Duration::from_secs(2);

#[derive(Debug, Clone, Serialize)]
pub struct Endpoint {
    pub source: String,
    pub browser_ws_url: String,
    pub http_version_url: Option<String>,
    pub version: Option<VersionInfo>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct VersionInfo {
    #[serde(rename = "Browser", default)]
    pub browser: Option<String>,
    #[serde(rename = "Protocol-Version", default)]
    pub protocol_version: Option<String>,
    #[serde(rename = "User-Agent", default)]
    pub user_agent: Option<String>,
    #[serde(rename = "V8-Version", default)]
    pub v8_version: Option<String>,
    #[serde(rename = "webSocketDebuggerUrl", default)]
    pub web_socket_debugger_url: Option<String>,
}

/// Resolve only explicitly supplied endpoints (args + env vars). Returns None
/// when nothing explicit was set; the caller decides whether to fall back to
/// profile-scan discovery.
pub async fn resolve_explicit_endpoint(
    browser_ws_url: Option<&str>,
    http_url: Option<&str>,
) -> anyhow::Result<Option<Endpoint>> {
    if let Some(url) = non_empty(browser_ws_url) {
        return Ok(Some(Endpoint {
            source: "argument".into(),
            browser_ws_url: url.into(),
            http_version_url: None,
            version: None,
        }));
    }
    if let Some(url) = non_empty(http_url) {
        return Ok(Some(endpoint_from_http_url(url, "argument").await?));
    }
    if let Some(url) = env_var("SOCAI_CDP_WS") {
        return Ok(Some(Endpoint {
            source: "SOCAI_CDP_WS".into(),
            browser_ws_url: url,
            http_version_url: None,
            version: None,
        }));
    }
    if let Some(url) = env_var("SOCAI_CDP_URL") {
        return Ok(Some(endpoint_from_http_url(&url, "SOCAI_CDP_URL").await?));
    }
    Ok(None)
}

/// One-shot discovery: explicit endpoints first, then `DevToolsActivePort` /
/// `/json/version` probes on the standard Chrome profile roots and ports.
pub async fn discover_existing_chrome_endpoint() -> anyhow::Result<Option<Endpoint>> {
    if let Some(endpoint) = resolve_explicit_endpoint(None, None).await? {
        return Ok(Some(endpoint));
    }
    for profile in chrome_profile_roots() {
        if let Some(endpoint) = endpoint_from_active_port(&profile).await {
            return Ok(Some(endpoint));
        }
    }
    for port in DEFAULT_DEVTOOLS_PORTS {
        let url = format!("http://127.0.0.1:{port}");
        if let Ok(endpoint) = endpoint_from_http_url(&url, &format!("port:{port}")).await {
            return Ok(Some(endpoint));
        }
    }
    Ok(None)
}

/// Poll `discover_existing_chrome_endpoint` until it succeeds or `timeout`
/// elapses. Intended for the first-run path where we opened chrome://inspect
/// and now wait for the user to approve remote debugging.
pub async fn wait_for_existing_chrome_endpoint(
    timeout: Duration,
    poll: Duration,
) -> anyhow::Result<Option<Endpoint>> {
    let deadline = Instant::now() + timeout.max(Duration::from_millis(100));
    let step = poll.max(Duration::from_millis(100));
    loop {
        if let Some(endpoint) = discover_existing_chrome_endpoint().await? {
            return Ok(Some(endpoint));
        }
        if Instant::now() >= deadline {
            return Ok(None);
        }
        tokio::time::sleep(step).await;
    }
}

/// Open the chrome://inspect remote-debugging page in the user's default
/// Chrome. Best-effort; failures are swallowed so first-run UX still degrades
/// gracefully to a printed instruction.
pub fn open_remote_debugging_page() {
    #[cfg(target_os = "macos")]
    {
        let _ = std::process::Command::new("open")
            .args(["-a", "Google Chrome", INSPECT_URL])
            .status();
    }
    #[cfg(target_os = "linux")]
    {
        let _ = std::process::Command::new("xdg-open")
            .arg(INSPECT_URL)
            .status();
    }
    #[cfg(target_os = "windows")]
    {
        let _ = std::process::Command::new("cmd")
            .args(["/C", "start", "", INSPECT_URL])
            .status();
    }
}

async fn endpoint_from_http_url(url: &str, source: &str) -> anyhow::Result<Endpoint> {
    let version_url = format!("{}/json/version", url.trim_end_matches('/'));
    let version: VersionInfo = get_json(&version_url).await?;
    let ws = version
        .web_socket_debugger_url
        .clone()
        .ok_or_else(|| anyhow::anyhow!("/json/version missing webSocketDebuggerUrl"))?;
    Ok(Endpoint {
        source: source.into(),
        browser_ws_url: ws,
        http_version_url: Some(version_url),
        version: Some(version),
    })
}

async fn endpoint_from_active_port(profile: &Path) -> Option<Endpoint> {
    let marker = profile.join("DevToolsActivePort");
    let contents = fs::read_to_string(&marker).ok()?;
    let mut lines = contents.lines();
    let port: u16 = lines.next()?.trim().parse().ok()?;
    let ws_path = lines.next().map(str::trim).unwrap_or("").to_string();

    // Prefer richer info via HTTP /json/version. Fall back to constructing the
    // ws URL ourselves if the HTTP endpoint refuses but DevToolsActivePort
    // gave us the path.
    let http_url = format!("http://127.0.0.1:{port}");
    let source = format!("active_port:{}", marker.display());
    if let Ok(endpoint) = endpoint_from_http_url(&http_url, &source).await {
        return Some(endpoint);
    }
    if ws_path.is_empty() {
        return None;
    }
    Some(Endpoint {
        source,
        browser_ws_url: format!("ws://127.0.0.1:{port}{ws_path}"),
        http_version_url: None,
        version: None,
    })
}

fn chrome_profile_roots() -> Vec<PathBuf> {
    if let Some(override_path) = env_var("SOCAI_CHROME_USER_DATA_DIR") {
        return vec![PathBuf::from(shellexpand(&override_path))];
    }
    let Some(home) = home_dir() else {
        return Vec::new();
    };

    #[cfg(target_os = "macos")]
    let candidates: &[&str] = &[
        "Library/Application Support/Google/Chrome",
        "Library/Application Support/Comet",
        "Library/Application Support/Arc/User Data",
        "Library/Application Support/Microsoft Edge",
        "Library/Application Support/BraveSoftware/Brave-Browser",
    ];

    #[cfg(target_os = "linux")]
    let candidates: &[&str] = &[
        ".config/google-chrome",
        ".config/chromium",
        ".config/microsoft-edge",
        ".config/BraveSoftware/Brave-Browser",
    ];

    #[cfg(target_os = "windows")]
    let candidates: &[&str] = &[
        "AppData/Local/Google/Chrome/User Data",
        "AppData/Local/Microsoft/Edge/User Data",
        "AppData/Local/BraveSoftware/Brave-Browser/User Data",
    ];

    candidates.iter().map(|c| home.join(c)).collect()
}

fn home_dir() -> Option<PathBuf> {
    if let Some(home) = env_var("HOME") {
        return Some(PathBuf::from(home));
    }
    #[cfg(target_os = "windows")]
    {
        if let Some(profile) = env_var("USERPROFILE") {
            return Some(PathBuf::from(profile));
        }
        if let (Some(drive), Some(path)) = (env_var("HOMEDRIVE"), env_var("HOMEPATH")) {
            return Some(PathBuf::from(format!("{drive}{path}")));
        }
    }
    None
}

fn env_var(key: &str) -> Option<String> {
    env::var(key).ok().filter(|v| !v.is_empty())
}

fn non_empty(s: Option<&str>) -> Option<&str> {
    s.filter(|v| !v.is_empty())
}

fn shellexpand(s: &str) -> String {
    if let Some(rest) = s.strip_prefix("~/") {
        if let Some(home) = home_dir() {
            return home.join(rest).to_string_lossy().into_owned();
        }
    }
    s.to_string()
}

async fn get_json<T: for<'de> serde::Deserialize<'de>>(url: &str) -> anyhow::Result<T> {
    let resp = reqwest::Client::builder()
        .timeout(HTTP_TIMEOUT)
        .build()?
        .get(url)
        .send()
        .await?
        .error_for_status()?;
    Ok(resp.json().await?)
}
